llama.cpp

```diff
@@ -11061,8 +11117,8 @@ static size_t llama_output_reserve(llama_context & lctx, size_t n_outputs) {
     const auto n_embd  = hparams.n_embd;
 
     // TODO: use a per-batch flag for logits presence instead
-    const bool has_logits = cparams.causal_attn;
-    const bool has_embd   = cparams.embeddings && (hparams.causal_attn || cparams.pooling_type == LLAMA_POOLING_TYPE_NONE);
+    const bool has_logits = !cparams.embeddings;
+    const bool has_embd   =  cparams.embeddings;
```

Comment:
Note that `lctx.embd` is not used by all pooled embeddings types, it's really only used with `LLAMA_POOLING_TYPE_NONE`.

(This is all done near the end of `llama_decode_internal` in a `switch` statement)

So maybe the condition for `has_embd` could be `cparams.embeddings && cparams.pooling_type == LLAMA_POOLING_TYPE_NONE`?

(See also the other places where `hparams.causal_attn` was used to understand the assumptions that stem from it, to check if they need to be modified)

Reply:
Ah, that makes sense. I put in the `pooling_type` check you suggested in there. I also changed the `inp_out_ids` calculation to rely on `cparams.embeddings` rather than `hparams.causal_attn`.